File: code/backend/src/main/java/com/taskflow/modules/analytics/service/AnalyticsService.java
package com.taskflow.modules.analytics.service;

import com.taskflow.modules.analytics.dto.ProductivityOverviewDto;

import java.util.UUID;

/**
 * Domain Service interface for calculating productivity metrics, task completion trends, and status breakdowns.
 */
public interface AnalyticsService {

    /**
     * Calculates productivity overview metrics.
     *
     * @param userId      UUID identifier of requesting user
     * @param workspaceId optional workspace scope
     * @param projectId   optional project scope
     * @param period      DAILY, WEEKLY, or MONTHLY aggregation window
     * @return ProductivityOverviewDto payload
     */
    ProductivityOverviewDto getProductivityOverview(UUID userId, UUID workspaceId, UUID projectId, String period);
}

File: code/frontend/src/features/analytics/services/analytics-service.ts
import { apiClient } from '@/lib/api-client';
import type { ApiResponse } from '@/features/auth/types';
import type { AnalyticsPeriod, ProductivityOverviewDto } from '../types';

export const analyticsService = {
  getProductivityOverview: async (
    period: AnalyticsPeriod = 'WEEKLY',
    workspaceId?: string,
    projectId?: string
  ): Promise<ProductivityOverviewDto> => {
    const res = await apiClient.get<ApiResponse<ProductivityOverviewDto>>('/analytics/overview', {
      params: { period, workspaceId, projectId },
    });
    return res.data.data;
  },
};
